File: services/document-indexer/gpu_vorrang.py
"""
Vorrang fuer den Nutzer an der einen GPU (Plan 023, Fund vom 22.08.2026).

Das Geraet hat eine GPU. Der Chat serialisiert seine Modellaufrufe im Backend
ueber `services/flows/gpuQueue.js`, und dessen Kopfzeile nennt sich selbst "die
EINE Sperre fuer alle lokalen Modell-Aufrufe". Sie ist es nicht: der Indexer
ist ein eigener Prozess in einem eigenen Container und ruft Ollama direkt auf.
Er kann diese Sperre nicht nehmen.

**Was daraus wurde, gemessen am 22.08.2026 auf dem Orin.** Der Indexer reichert
im Hintergrund an und laedt dafuer `qwen3:14b` (14 GB). Der Chat rechnet mit
dem 27B-Modell (22 GB). Zusammen passen sie nicht in das Budget, also wirft
Ollama das jeweils andere Modell heraus. In `llm_model_switches` stehen fuer
vierzig Minuten 35 Zeilen `auto_unload_ollama_keepalive` im Wechsel und neun
Ladevorgaenge mit 11 827 bis 60 066 Millisekunden. Der Nutzer wartet also bei
jeder Chat-Runde eine halbe bis eine ganze Minute darauf, dass ein Modell
zurueck in den Speicher kommt, das kurz zuvor schon dort war.

**Die Loesung ist eine Frist, kein Schalter.** Das Backend meldet "ich habe die
GPU, halte dich bis T zurueck" und erneuert das, solange es rechnet. Faellt das
Backend aus, laeuft die Frist ab und der Indexer arbeitet weiter. Ein Schalter
haette den Indexer nach einem Absturz fuer immer stillgelegt.

Zurueckgehalten wird nur der START eines Modellaufrufs. Ein laufender Aufruf
wird nicht abgebrochen: das Modell rechnet ohnehin zu Ende, und ein halbes
Ergebnis waere schlechter als gar keins. Dieselbe Ueberlegung steht bei
`ai_services.analyze_document`.
"""

import os
import threading
import time

# Laenger als eine Frist, die das Backend setzt, wartet der Indexer nie. Auch
# nicht, wenn das Backend eine unsinnig grosse Zahl schickt: eine Zusage, die
# der Indexer eine Viertelstunde lang stillegt, waere ein Ausfall, kein
# Vorrang.
HOECHSTFRIST_S = float(os.getenv('GPU_VORRANG_HOECHSTFRIST_S', '120'))

_sperre = threading.Lock()
_frist_bis = 0.0


def melde_belegt(sekunden: float) -> float:
    """
    Setzt die Frist auf `jetzt + sekunden`.

    Eine bereits laufende, laengere Frist wird nicht verkuerzt, ausser die
    Meldung lautet ausdruecklich Null: das ist das Freigeben am Ende.

    :returns: die Sekunden, die tatsaechlich gelten.
    """
    global _frist_bis
    with _sperre:
        if sekunden <= 0:
            _frist_bis = 0.0
            return 0.0
        gilt = min(float(sekunden), HOECHSTFRIST_S)
        neu = time.monotonic() + gilt
        if neu > _frist_bis:
            _frist_bis = neu
        return max(0.0, _frist_bis - time.monotonic())


def gpu_belegt() -> bool:
    """True, solange das Backend die GPU fuer einen Nutzerlauf haelt."""
    with _sperre:
        return time.monotonic() < _frist_bis


def restsekunden() -> float:
    """Wie lange die Frist noch laeuft. Null, wenn sie abgelaufen ist."""
    with _sperre:
        return max(0.0, _frist_bis - time.monotonic())


def zuruecksetzen() -> None:
    """Nur fuer Tests."""
    global _frist_bis
    with _sperre:
        _frist_bis = 0.0

File: apps/dashboard-backend/src/services/flows/gpuVorrang.js
/**
 * Dem Indexer sagen, dass der Nutzer gerade die GPU braucht (Fund 22.08.2026).
 *
 * `gpuQueue.js` nennt sich in seiner Kopfzeile "die EINE Sperre fuer alle
 * lokalen Modell-Aufrufe". Das stimmte fuer alles, was IN DIESEM PROZESS
 * laeuft. Der Document-Indexer ist ein eigener Container, ruft Ollama direkt
 * auf und kann diesen Mutex nicht nehmen.
 *
 * **Gemessen auf dem Orin am 22.08.2026.** Der Indexer reichert im Hintergrund
 * an und laedt dafuer `qwen3:14b` (14 GB), der Chat rechnet mit dem
 * 27B-Modell (22 GB). Beide zusammen passen nicht in das Budget, also wirft
 * Ollama abwechselnd das jeweils andere heraus. In `llm_model_switches` stehen
 * fuer vierzig Minuten 35 Zeilen `auto_unload_ollama_keepalive` und neun
 * Ladevorgaenge zwischen 11 827 und 60 066 Millisekunden. Der Nutzer wartet
 * also bei fast jeder Chat-Runde eine halbe bis eine ganze Minute auf ein
 * Modell, das kurz zuvor schon im Speicher war.
 *
 * **Eine Frist, kein Schalter.** Gemeldet wird "halte dich bis T zurueck", und
 * solange gerechnet wird, wird die Frist erneuert. Stirbt das Backend mitten
 * im Lauf, laeuft die Frist ab und der Indexer arbeitet von selbst weiter. Ein
 * Schalter haette ihn nach einem Absturz fuer immer stillgelegt.
 *
 * **Nie im Weg.** Jeder Aufruf ist abgeschickt und vergessen, mit kurzem
 * Zeitlimit. Ein Indexer, der nicht antwortet, darf den Chat nicht bremsen —
 * das waere derselbe Fehler in der anderen Richtung.
 */

// Bewusst `fetch` und nicht `axios`: die Meldung faehrt sonst auf derselben
// Attrappe wie die Modell-Aufrufe. In `runFlow.test.js` stand danach die
// Vorrang-Meldung als `axios.post.mock.calls[0]` vor dem Modell-Aufruf, und 21
// Tests prueften plotzlich den falschen Aufruf. Ein Nebenweg, der fremde Tests
// verschiebt, ist der falsche Nebenweg.
const logger = require('../../utils/logger');

const INDEXER_URL = process.env.DOCUMENT_INDEXER_URL || 'http://document-indexer:9102';
// Wie lange eine einzelne Meldung gilt. Kurz genug, dass eine vergessene
// Freigabe den Indexer nicht lange aufhaelt, lang genug, dass zwischen zwei
// Erneuerungen keine Luecke entsteht.
const FRIST_S = parseInt(process.env.GPU_VORRANG_FRIST_S || '30', 10);
const ERNEUERN_MS = Math.max(1000, Math.round((FRIST_S * 1000) / 3));
const ZEITLIMIT_MS = parseInt(process.env.GPU_VORRANG_ZEITLIMIT_MS || '1500', 10);
const AN = process.env.GPU_VORRANG !== 'false';

let meldungenFehlgeschlagen = 0;

async function melde(sekunden) {
  if (!AN) {return;}
  try {
    await fetch(`${INDEXER_URL}/gpu/vorrang`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ sekunden }),
      signal: AbortSignal.timeout(ZEITLIMIT_MS),
    });
    meldungenFehlgeschlagen = 0;
  } catch (err) {
    // Einmal laut, danach still. Ein abgeschalteter Indexer soll nicht bei
    // jedem Chat-Zug eine Zeile schreiben.
    //
    // `warn` und nicht `debug`: zwanzig Testdateien legen den Logger als
    // `{ info, warn, error }` an, und ein Aufruf von `debug` liess dort acht
    // fremde Tests scheitern. Eine Nebensache darf nicht die Stelle sein, an
    // der etwas bricht.
    if (meldungenFehlgeschlagen === 0) {
      logger.warn(`[GPU-Vorrang] Indexer nicht erreichbar: ${err.message}`);
    }
    meldungenFehlgeschlagen += 1;
  }
}

/**
 * Haelt den Vorrang, solange `fn` laeuft.
 *
 * @param {() => Promise<T>} fn
 * @returns {Promise<T>}
 * @template T
 */
async function mitVorrang(fn) {
  if (!AN) {return fn();}
  melde(FRIST_S);
  const takt = setInterval(() => melde(FRIST_S), ERNEUERN_MS);
  // Der Takt darf einen Neustart des Prozesses nicht aufhalten.
  if (typeof takt.unref === 'function') {takt.unref();}
  try {
    return await fn();
  } finally {
    clearInterval(takt);
    melde(0);
  }
}

module.exports = { mitVorrang, _melde: melde, _FRIST_S: FRIST_S };
